Linux CUDA support policy (#8150)
circleci/circleci-docs: jekyll/_cci2/linux-cuda-images-support-policy.adoc

---
contentTags:
  platform:
  - Cloud
  - Server v4.x
  - Server v3.x
---
= CircleCI Linux CUDA images support policy
:page-layout: classic-docs
:page-liquid:
:page-description: CircleCI Linux CUDA images release, update, and deprecation policy
:icons: font
:toc: macro
:toc-title:

[#overview]
== Overview

> This document outlines the xref:#[CircleCI Linux CUDA image] release, update, and deprecation policy and applies to our Linux CUDA Images.

Author: @JalexChen Is there an existing page in the docs that we can link to on here?

Reviewer: This should be it I think: https://circleci.com/docs/using-gpu/#linux-gpu

Reviewer 2: i think the link @rosieyohannan provided would be it, but i also think that we should be adding a CUDA page to the developer hub. The list of all images in `using-gpu` redirects [here](https://circleci.com/developer/images?imageType=machine), but none of the images for CUDA are actually listed there, and i think that can make things extremely confusing

Author: We can add the link to the dev hub page when it's ready 👍

> This document outlines the xref:using-gpu#linux-gpu[CircleCI Linux CUDA image] release, update, and deprecation policy and applies to our Linux CUDA Images.

Reviewer: ```suggestion
This document outlines the xref:using-gpu#linux-gpu[CircleCI Linux CUDA image] release, update, and deprecation policy and applies to our Linux CUDA images.
```

[#release-policy]
== Release policy

The CircleCI CUDA images are based on our Linux VM machine images for the purpose of providing an image that can leverage the additional processing capabilities of CircleCI’s GPU executor. 

We aim to support the three most recent minor versions (or three most recent even numbered minor versions, if minor version >= `6`) of the two most recent and supported major CUDA releases. We closely track Nvidia’s own release cycle for our releases and aim to release these within a week of the CUDA release. 

Please note that this is not an SLA (service level agreement). We can not, and do not, provide an official SLA turnaround time for new CUDA images.

[#tagging]
== Tagging

> The following are the types of tags available for the CircleCI Linux CUDA images: 

Author: @JalexChen I rephrased this section a bit. Let me know what you think

> The following are the types of tags available for the CircleCI Linux CUDA images: 

Reviewer: ```suggestion
The following tags are available for the CircleCI Linux CUDA images: 
```

* `linux-cuda-<CUDA-version>:<year.month.patch>`
** For example, `linux-cuda-11:2023.02.1`

> ** The “default” slugs will simply remove the tag if users want to keep up-to-date:

Author: @JalexChen I think this needs a bit of clarification or rephrasing. Do we mean to say that the default slug is just the slug without `year.month.patch` ? Which image would that point to? Am I correct in saying that "default" can change without notice?

This is what we have for Linux VM, let me know if it's the same: https://github.com/circleci/circleci-docs/pull/8011/files#diff-788abd596f1e0657f481d530f0bca4d7c3dce21e971ed4e8389aa6c21f711a47R47

Reviewer 2: So, the default tags are meant to cover off on a period of ~2 years, based on CUDA release history, so internally, we would keep the dated tags (up to three, just so there is overlap), but we ultimately want users to _only_  use the default tags, which i think actually need to look like this: `linux-cuda-11:default` or `linux-cuda-12:default`.

what are your thoughts on removing the references to dated tags?

the default tag will always point to the latest version of the image on a rolling basis and no, it will change, but with notice, the idea is that workflows are uninterrupted. there are three components to this:
- every minor/patch release within the major CUDA version is compatible, and we will include up to 3 of the past versions
- there will be an edge image that is created first
- a discuss post will be made letting users know of an update to the versions in the tag

Author: @JalexChen I think my preference is to be consistent with the other policy docs, and since this is also a reference type of document we should be detailed and thorough. So I'd still keep the dated tags on here, but would add a recommendation to readers about figuring out which tags work best for their team in the usage guide (e.g. https://circleci.com/docs/using-gpu/).

> *** For example, `linux-cuda-11` or `linux-cuda-12`

Reviewer 2: `linux-cuda-11:default` or `linux-cuda-12:default`

Reviewer 2: @kbatuigas i think this makes the most sense

* `linux-cuda-<CUDA-version>:edge`
** This will be available a week prior to general releases. These images incorporate any new changes for customers who want the latest version as fast as possible.

[#critical-cve-patches]
== Critical CVE patches

> When critical CVEs are disclosed that affect the versions of the operating system or software stack in either Linux or Windows images, we will investigate the impact that this has on our images being used within the CircleCI execution environment. 

Reviewer: ```suggestion
When critical CVEs are disclosed that affect the versions of the operating system or software stack in either Linux or Windows images, we will investigate the impact that this has on our images being used across CircleCI execution environments. 
```

In most cases, due to the ephemeral and isolated nature of the environment, it is not necessary to patch these images. We will always communicate our stance on these disclosures via our link:https://discuss.circleci.com/[Discuss Forum].

[#bug-reports-issues-and-prs]
== Bug reports, issues, and PRs

> Please file a link:https://support.circleci.com/hc/en-us/requests/new[support ticket] with CircleCI Support for any issues or bugs found with the Linux CUDA images. Our support team will be able to escalate issues internally to the correct engineering team and provide updates on the issue.

Reviewer: ```suggestion
Please file a link:https://support.circleci.com/hc/en-us/requests/new[support ticket] with CircleCI Support for any issues or bugs found with the CircleCI Linux CUDA images. Our support team will be able to escalate issues internally to the correct engineering team and provide updates on the issue.
```

[#image-lifespan-eol]
== Image lifespan / EOL

Image lifespan will generally follow Nvidia’s CUDA release cycle. 

Once a new major version of CUDA is released, we will begin the deprecation process of the oldest major version that we support and schedule it to be removed. This allows us to maintain two major versions of CUDA effectively.

> Example: When CUDA 12.0 is released:

Reviewer: Maybe use a heading to avoid double ::?

Reviewer 2: good idea. i made the change in the [discuss post](https://discuss.circleci.com/t/linux-cuda-deprecation-and-image-policy/48568) as well

[.table.table-striped]
[cols=2*, options="header", stripes=even]
|===
| Version
| Support

| CUDA 12
| `linux-cuda-12` image added

| CUDA 11
| No change

| CUDA 10
| `linux-cuda-10` image deprecated
|===

Example: When CUDA 11.8 is released:

[.table.table-striped]
[cols=2*, options="header", stripes=even]
|===
| Version
| Support

| CUDA 11.8
| Added to `linux-cuda-11`

| CUDA 11.6
| Remains in the image

| CUDA 11.4
| Remains in the image; becomes “oldest”

| CUDA 11.2
| Removed from `linux-cuda-11`
|===

Once a new even numbered minor version is released, that version will be added and the oldest existing version will be removed, provided that more than 3 minor versions exist.

> When an image is selected for deprecation and removal, we will create an announcement on our Discuss forum, along with reaching out via email to developers who have requested one of the deprecated images in their recent jobs. We will also plan brownouts to ensure users are aware of the approaching removal of deprecated images.

Reviewer: ```suggestion
When an image is selected for deprecation and removal, we will create an announcement on our Discuss forum, along with reaching out via email to developers who have requested one of the deprecated images in their recent jobs. We will also plan brownouts to help ensure you are aware of the approaching removal of deprecated images.
```

Reviewer: Are there in-app banners? @JalexChen

Reviewer 2: @rosieyohannan - yes, the idea is to have in-app banners right before/during brownouts. i'll need to coordinate with execution/fleets on this
